### PR TITLE
Make Object and derived classes use SharedPtr with the Original mapping

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR changes ObjectPtr to be a SharedPtr<Object> with the Original mapping. It's the first step in making the same Ice::Object the common Object base class for the Original and New mappings.

You'll notice:
 - there are few C++ updates, especially test updates, which is good as it shows this update is largely source compatible
 - the updates are mostly related to "this" sharing
 - there was no update to PHP and Ruby, and minimal updates to Python
 - unfortunately Objective-C required more updates as it was overwriting IceUtil::Shared __incRef and __decRef.

### CODE REVIEW

File: objective-c/src/Ice/Object.mm [hunk 1]
```diff
@@ -77,7 +74,16 @@ virtual void __decRef()
 {
 public:
 
-    BlobjectI(ICEBlobject*);
+    BlobjectI(ICEBlobject* servant) : _blobject(servant), _target(servant)
```
Comment (pepone) [inline objective-c/src/Ice/Object.mm:77]: I know you didn't change this, but why do we need two pointers to the same object here?

File: objective-c/src/Ice/PropertiesI.mm [hunk 2]
```diff
@@ -270,21 +270,21 @@ -(void) load:(NSString*)file
...
-typedef IceUtil::Handle<UpdateCallbackI> UpdateCallbackIPtr;
+using UpdateCallbackIPtr = Ice::SharedPtr<UpdateCallbackI>;
 
 }
 
 @implementation ICEPropertiesAdminUpdateCallback
 @end
 
-#define NATIVEPROPERTIESADMIN dynamic_cast<Ice::NativePropertiesAdmin*>(object_)
+#define NATIVEPROPERTIESADMIN std::dynamic_pointer_cast<Ice::NativePropertiesAdmin>(object_)
```
Comment (pepone) [inline objective-c/src/Ice/PropertiesI.mm:280]: I think we should remove this macro, which its only purpose seems to be to save some typing.

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]
Looks good

File: objective-c/src/Ice/Object.mm (hunk 1 above)
Comment (bernardnormier) [reply objective-c/src/Ice/Object.mm:77]: You're right, this is unnecessary. I fixed it in #1667.

File: objective-c/src/Ice/PropertiesI.mm (hunk 2 above)
Comment (bernardnormier) [reply objective-c/src/Ice/PropertiesI.mm:280]: I agree, for a future PR. Note that we have several of these macros.

**Pull request merged**